File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = 4;
        char[][] a = new char[4][8];
        for(int i=0; i<n; i++){
            a[i]  = sc.next().toCharArray();
        }
        int k = sc.nextInt(); // 반복 횟수 입력받음

        while(k>0){
            k--;
            int no = sc.nextInt()-1; // 톱니바퀴 번호 입력
            int dir = sc.nextInt(); // 회전 방향 입력

            // 톱니바퀴들은 동시에 회전을 해야되기 때문에 각 톱니바퀴들의 회전 방향을 저장할 배열 선언
            int[] d = new int[n]; // 톱니바퀴 수 만큼 배열 생성
            d[no] = dir;

            // 왼쪽 톱니바퀴들 부터 확인
            for(int i=no-1; i>=0; i--){
                if(a[i][2]!=a[i+1][6]){
                    d[i] = -d[i+1];
                }else{
                    break;
                }
            }

            // 오른쪽 톱니바퀴들 확인
            for(int i=no+1; i<n; i++){
                if(a[i-1][2]!=a[i][6]){
                    d[i] = -d[i-1];
                }else{
                    break;
                }
            }

            // 저장한 회전 방향 정보들을 가지고 톱니바퀴를 회전시킴
            for(int i=0; i<n; i++){
                if(d[i]==0) continue;
                if(d[i]==1){
                    // 시계 방향 회전 (오른쪽으로 한칸씩 이동)
                    char temp = a[i][7];
                    for(int j=7; j>=1; j--){
                        a[i][j] = a[i][j-1];
                    }
                    a[i][0] = temp;

                }else if(d[i]==-1){
                    // 반시계 방향 회전 (왼쪽으로 한칸씩 이동)
                    char temp = a[i][0];
                    for(int j=0; j<7; j++){
                        a[i][j] = a[i][j+1];
                    }
                    a[i][7] = temp;
                }
            }
        }

        int ans = 0;
        if(a[0][0]=='1') ans+=1;
        if(a[1][0]=='1') ans+=2;
        if(a[2][0]=='1') ans+=4;
        if(a[3][0]=='1') ans+=8;

        System.out.println(ans);

    }
}

File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.py
n = 4
a = [list(input()) for _ in range(n)]
k = int(input())

for _ in range(k):
    no, dir = map(int,input().split())
    no-=1 # 배열 인덱스는 0부터 시작이므로
    d = [0]*n # 각 톱니바퀴의 회전 방향을 저장하기 위한 배열
    d[no] = dir
    # 현재 톱니에서 왼쪽 톱니를 연쇄적으로 구함
    for i in range(no-1,-1,-1):
        if a[i][2] != a[i+1][6]:
            d[i] = -d[i+1] # 현재 톱니와 반대 방향으로 회전하는 정보를 저장
        else: # 하나라도 회전을 하지 않는 톱니가 나타나면 연쇄 회전은 중단
            break

    for i in range(no+1,n):
        if a[i-1][2] != a[i][6]:
            d[i] = -d[i-1]
        else:
            break

    for i in range(n): # 각 톱니바퀴마다 회전 방향을 토대로 회전 시키기
        if d[i] == 0: # 0은 회전하지 않는 것을 의미
            continue
        if d[i] == 1: # 시계방향 회전
            temp = a[i][7]
            for j in range(7,0,-1):
                a[i][j] = a[i][j-1]
            a[i][0] = temp
        elif d[i] ==-1: # 반시계방향 회전
            temp = a[i][0]
            for j in range(0,7):
                a[i][j] = a[i][j+1]
            a[i][7] = temp


ans = 0
if a[0][0]=='1':
    ans+=1
if a[1][0]=='1':
    ans+=2
if a[2][0]=='1':
    ans+=4
if a[3][0]=='1':
    ans+=8

print(ans)

